src/librustc_mir/interpret/validity.rs

```diff
@@ -554,12 +560,16 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
             // generators and closures.
             ty::Closure(def_id, _) | ty::Generator(def_id, _, _) => {
                 if let Some(node_id) = self.tcx.hir.as_local_node_id(def_id) {
-                    let freevar = self.tcx.with_freevars(node_id, |fv| fv[field]);
-                    PathElem::ClosureVar(self.tcx.hir.name(freevar.var_id()))
-                } else {
-                    // The closure is not local, so we cannot get the name
-                    PathElem::ClosureVar(Symbol::intern(&field.to_string()))
+                    if let Some(freevar) = self.tcx.with_freevars(
+                        node_id,
+                        |fv| fv.get(field).map(|field| *field))
+                    {
+                        return PathElem::ClosureVar(self.tcx.hir.name(freevar.var_id()));
+                    }
                 }
+                // The closure is not local, or the freevars don't match up (seen for a generator!),
```

Comment:
`tcx.mir_optimized(def_id).upvar_decls[upvar_index].debug_name` should do it?

Reply:
There is [no such method](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/?search=mir_optimized)?

Reply:
sorry, was on mobile

https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.optimized_mir

Reply:
What will that do if we donÄt have libstd MIR?

Reply:
uhm... I don't see how that can happen. How would you even see a libstd closure without its MIR being available?

Reply:
Fair enough.

However, for a generator, I am still seeing

> index out of bounds: the len is 0 but the index is 0

So I'll have to keep some kind of fallback.

Reply:
Done, and even added a test for it. ;)